File: app/src/main/java/com/auto/billiardnote/ui/home/draw/Ball.java
package com.auto.billiardnote.ui.home.draw;

import android.graphics.Canvas;
import android.graphics.Paint;

import com.auto.billiardnote.ui.home.draw.shape.Circle;

public class Ball extends Circle {

    private final Paint paint;
    private final int color;

    public Ball (float x, float y, float r, DrawingTool tool)
    {
        super(x, y, r);
        paint = DrawingTool.getPaint(tool);
        this.color = paint.getColor();
    }

    public int getColor() { return color; }

    public boolean touch_move(float x, float y) {
        float dx = super.abs(x, true);
        float dy = super.abs(x, false);
        float TOLERANCE = 4;
        if (dx >= TOLERANCE || dy >= TOLERANCE) {
            super.setCoordinate(Math.round(x), Math.round(y));
            return true;
        }
        return false;
    }

    public void draw(Canvas canvas) {
        super.draw(canvas, paint);
    }

    public void setClickListener() {
    }

    public boolean isWithin(float x, float y) {
        return super.isWithin(x, y);
    }
}

File: app/src/main/java/com/auto/billiardnote/ui/home/draw/Ball.kt
package com.auto.billiardnote.ui.home.draw

import android.graphics.Canvas
import android.graphics.Paint
import com.auto.billiardnote.ui.home.draw.shape.Circle
import kotlin.math.roundToInt

class Ball {
    private val paint: Paint?
    val color: Int
    private val circle: Circle

    constructor(ball: Ball) {
        var tmp: Paint? = null
        for (tool in DrawingTool.values()) {
            if (tool.isSame(ball.color)) {
                tmp = DrawingTool.Companion.getPaint(tool)
                break
            }
        }
        circle = Circle(ball.circle)
        paint = tmp
        color = paint!!.color
    }

    constructor(x: Float, y: Float, r: Float, tool: DrawingTool) {
        circle = Circle(x, y, r)
        paint = DrawingTool.Companion.getPaint(tool)
        color = paint.color
    }

    fun touch_move(x: Float, y: Float): Boolean {
        val dx = circle.abs(x, true)
        val dy = circle.abs(x, false)
        val tolerance = 4f
        if (dx >= tolerance || dy >= tolerance) {
            circle.setCoordinate(x.roundToInt().toFloat(), y.roundToInt().toFloat())
            return true
        }
        return false
    }

    fun draw(canvas: Canvas) {
        circle.draw(canvas, paint)
    }

    fun setClickListener() {}
    fun isWithin(x: Float, y: Float): Boolean {
        return circle.isWithin(x, y)
    }
}